docker/local-test/README.md

```diff
@@ -0,0 +1,117 @@
+run our tests from inside there, and wow zippy. Debugging now feasible.
+
+## How
+
+1. The image is built with all the contents necessary to install and run PyCDS
+and its tests. 
+
+1. But since we want to run our own, local test code, we can't install PyCDS 
+from a repo. Instead we install from our local codebase when the container is 
+started.
+
+1. To facilitate this, we set up a working diretory (WORKDIR) in the image
+called `/codebase`. 
+
+1. When we run the image, we must mount our local codebase to `/codebase`.
+(See Run image).
+
+1. When the container starts (image runs), the script 
+`entrypoint.sh` installs the local version of PyCDS (in development mode `-e`). 
+It also sets up 
+and `su`s a non-root user, `test`, because PostgreSQL refuses -- sensibly --
+to run as the root user, which is what we are up to this point.
+
+1. Because we have mounted our codebase to the
+container, when we make changes to it (outside the container), those changes
+are available inside the container, and vice-versa. Therefore we can use all
+our local tools outside the container as normal (which is a shedload easier
+than trying to install your IDE inside the container :) ).
+
+1. The vice-versa has a downside, which is that runs of the tests leave
+behind a set "orphaned" pytest caches which will cause the next
+run of the image to fail if they are not cleaned up first with `py3clean`.
+We don't, however mount the codebase read-only because we might want 
+some effects of the test runs to be written to our external filesystem 
+(e.g., redirected output).
+
+## Notes and caveats
+
+1. Writing to a mounted volume from inside a docker container involves some
+tricky permissions logic that I don't fully understand yet. Known:
+    - If the user inside the container has the same user id 
+    (numeric, e.g., 1000) as the owner of the mounted file or directory outside 
+    the container, then it is possible to write to the mounted volume 
+    (e.g., to redirect output from a test run to a file). 
+    - If the user id's differ, a permissions error is raised and the write 
+    fails.
+    - Default user id (of the first user) on a Linux system is 1000.
+    My own user id (rglover) is 1000. Hence the setting of user id 1000
+    in `entrypoint.sh`. If your user id is not 1000, you will need to change
+    this if you wish to write content from inside the container.
+    - This is a hack and should be cleaned up so it works
+    generally. That will require some research into Docker's mechanisms for
+    mapping user ids between a container and its run environment, not for the
+    fainthearted. 
+
+1. As noted above, running tests in the test container in read/write mode 
+leaves problematic pycache junk behind in the host filesystem. 
+This can be cleaned up by running `py3clean`.
+
+## Pull image
+
+The GitHub Action docker-publish automatically builds the image.
+Pull it from Dockerhub:
+
+```
+docker pull pcic/pycds:local-test
```

Comment:
Is there any reason for us to keep the image produced by `dev.Dockerfile`? If not should we consider having this `local-test` image follow the typical docker naming convention? While the tag name works for now, the amount of overwriting that will be done may start to cause confusion.

Reply:
That makes total sense to me. I'll update the action and you can correct me on any mistakes I make.

Reply:
Actually I just realized, after *reading my own documentation* :brain: , that this dockerfile (a) still serves a useful purpose, and (b) that purpose is distinct from the purpose of the local-test dockerfile. So I will shuttle things around but we will still have two docker images. There's a naming issue here ... how do we resolve it?

Reply:
I was actually going to address the overall image name. If I understand correctly, neither of these images are actually used to create an instance of the application (in the way that the p2a-v2-frontend image is used, for example). If that is in fact the case, I think using the pycds name is a bit disingenuous and a bit confusing (hence why I am confused haha). Therefore we can have two separate dockerhub repos for the two different images being created and they can both use the naming standard. Your image name suggestions sound great to me! I would just recommend that you move dev.dockerfile to its own folder in docker/.

Reply:
Can the two images be given different names? If so, I'd suggest `pcic/pycds-local-pytest` for the one, and `pcic/pycds-test-db` for the other, and let tagging be as usual. Is that possible?

Reply:
That is possible, they would just live in different places on dockerhub.

Reply:
I think we are on exactly the same wavelength here. 

You do understand correctly; there is no application _per se_; this is a library package, primarily used by other applications. (Although I suppose the scripts in this package constitute an application. We could install it in a Dockerfile so that the scripts could be run from there, but that seems unnecessary.)

Also agree on moving the dev.Dockerfile into `docker/`.

I don't know how to establish the naming conventions we both agree on here, but I will do the rest of the work and let you tweak the docker-publish action accordingly.

Reply:
This [line](https://github.com/pacificclimate/pycds/blob/i64-wa-matviews/.github/workflows/docker-publish.yml#L24) determines the image name. So you would place `pcic/pycds-test-db` there. `tag_with_ref` should be set to `true` and the tag naming will be taken care of by the action.